src/backend/cpu/convolve.cpp

```diff
@@ -124,5 +134,169 @@ INSTANTIATE(ushort, float)
+
+    dim_t outputWidth =
+        1 +
+        (sDims[0] + 2 * padding[0] - (((fDims[0] - 1) * dilation[0]) + 1)) /
+            stride[0];
+    dim_t outputHeight =
+        1 +
+        (sDims[1] + 2 * padding[1] - (((fDims[1] - 1) * dilation[1]) + 1)) /
+            stride[1];
+
+    const bool retCols    = false;
+    Array<accT> unwrapped = unwrap_dilated(
+        cast<accT>(signal), fDims[0], fDims[1], stride[0], stride[1],
+        padding[0], padding[1], dilation[0], dilation[1], retCols);
+
+    unwrapped  = reorder(unwrapped, dim4(1, 2, 0, 3));
+    dim4 uDims = unwrapped.dims();
+    unwrapped.modDims(dim4(uDims[0] * uDims[1], uDims[2] * uDims[3]));
+
+    vector<af_seq> flip_index(4);
+    af_seq s      = {(double)(fDims[0] - 1), 0, -1};
+    flip_index[0] = s;
+    s             = {(double)(fDims[1] - 1), 0, -1};
+    flip_index[1] = s;
+    flip_index[2] = af_span;
+    flip_index[3] = af_span;
+
+    Array<accT> collapsedFilter = createSubArray(filter, flip_index);
+    collapsedFilter.modDims(dim4(fDims[0] * fDims[1] * fDims[2], fDims[3]));
+
+    Array<accT> res =
+        matmul(unwrapped, collapsedFilter, AF_MAT_TRANS, AF_MAT_NONE);
+    res.modDims(dim4(outputWidth, outputHeight, signal.dims()[3],
+                     collapsedFilter.dims()[1]));
+    Array<T> out = cast<T>(reorder(res, dim4(0, 1, 3, 2)));
+
+    return out;
+}
+
+template <typename T, typename accT>
+Array<T> convolve2(Array<T> const &signal, Array<accT> const &filter,
+                   const dim4 stride, const dim4 padding, const dim4 dilation) {
+    Array<T> out = createEmptyArray<T>(dim4());
+    out = convolve2_unwrap<T, accT>(signal, filter, stride, padding, dilation);
+
+    return out;
+}
+
+#define INSTANTIATE(T, accT)                                                  \
+    template Array<T> convolve2<T, accT>(                                     \
+        Array<T> const &signal, Array<accT> const &filter, const dim4 stride, \
+        const dim4 padding, const dim4 dilation);
+
+INSTANTIATE(double, double)
+INSTANTIATE(float, float)
+INSTANTIATE(uint, float)
+INSTANTIATE(int, float)
+INSTANTIATE(uchar, float)
+INSTANTIATE(ushort, float)
+INSTANTIATE(short, float)
+#undef INSTANTIATE
+
+template <typename T, typename accT>
+Array<T> conv2DataGradient(const Array<T> &incoming_gradient,
+                           const Array<T> &original_signal,
+                           const Array<accT> &original_filter,
+                           const Array<T> &convolved_output, af::dim4 stride,
+                           af::dim4 padding, af::dim4 dilation) {
+    const dim4 cDims = incoming_gradient.dims();
+    const dim4 sDims = original_signal.dims();
+    const dim4 fDims = original_filter.dims();
+
+    vector<af_seq> flip_index(4);
+    af_seq s      = {(double)(fDims[0] - 1), 0, -1};
+    flip_index[0] = s;
+    s             = {(double)(fDims[1] - 1), 0, -1};
+    flip_index[1] = s;
+    flip_index[2] = af_span;
+    flip_index[3] = af_span;
```

Comment:
the flipping operation can be extracted out into a function like `genFlipIndex(std::array<af_seq, 4> &out, dim4 dims)` since it is being used in multiple locations. Also, std::array should be used instead of std::vector here.

which would turn the above section into something like
```
std::array<af_seq, 4> flip_index;
genFlipIndex(flip_index, fDims);
```